app/src/app/copilot/chat/bubble.tsx

```diff
@@ -0,0 +1,170 @@
+"use client";
+
+import { type Message } from "ai";
+import { cn } from "~/lib/utils";
+import { Markdown } from "../markdown/react-markdown";
+import { Avatar } from "~/components/ui/avatar";
+import { RateMessage } from "./rate-message";
+
+const verifiedSourcesText = "Verified Sources:";
+
+function getSources(message: Message) {
+  const sourcesIndex = message.content.indexOf(verifiedSourcesText);
+
+  if (sourcesIndex === -1 || message.role !== "assistant") {
+    return {
+      sources: [],
+      messageWithOutSources: message.content,
+      messageId: 0,
+    };
+  }
+
+  const idIndex = sourcesIndex - 9;
+
+  // Extract everything up to "Verified Sources:"
+  const messageWithOutSources = message.content.substring(0, idIndex);
+  const messageId = Number(
+    message.content.substring(idIndex, sourcesIndex).trim(),
+  );
+
+  // Extract the substring from "Verified Sources" to the end of the string
+  const verifiedSourcesSection = message.content.substring(sourcesIndex).trim();
+
+  // Split the section into lines and remove the first line
+  const [, ...sourcesLines] = verifiedSourcesSection.split("\n");
+
+  // Keep the URLs
+  const sources = sourcesLines.map((line) => {
+    // You can further process each line if necessary, for example, extract URLs from markdown links
+    const urlMatch = line.match(/\[.*?\]\((?<url>.*?)\)/);
+
+    if (urlMatch?.groups?.url) {
+      return urlMatch.groups.url;
+    }
+
+    return line.trim(); // This will return the URL part of the markdown link
+  });
+
+  return {
+    sources,
+    messageWithOutSources,
+    messageId,
+  };
+}
+
+function getDisplayNameFromURL(url: string) {
+  try {
+    const urlObj = new URL(url);
+    // Remove common subdomains
+    const hostname = urlObj.hostname.replace(/^www\./, "");
+
+    // Extract the second-level domain part
+    const parts = hostname.split(".");
+    const displayName = parts.length > 1 ? parts.at(-2)! : hostname;
+
+    // Capitalize the first letter
+    return displayName.charAt(0).toUpperCase() + displayName.slice(1);
+  } catch (e) {
+    console.error(e, url);
+    return url;
+  }
+}
+
+export default function Bubble({
+  message,
+  loading = false,
+}: {
+  message: Message;
+  loading?: boolean;
+}) {
+  const { sources, messageWithOutSources, messageId } = getSources(message);
+
+  return (
+    <div
+      key={message.id}
+      className={`flex flex-col items-center border-none ${
+        message.role === "user"
+          ? "bg-human-background text-human-foreground border-b border-gray-900/50"
+          : "bg-assistant-background text-assistant-foreground border-b border-gray-900/50"
```

Comment:
Non-Blocking: Add in css for assistant and human text blocks

Reply:
Solved by [e246a83](https://github.com/near-horizon/horizon/pull/72/commits/e246a83b27c259d97f8572b24ad6ddaae6c0cf21)

This should be revisited to improve styling as it looks off compared to the rest of the app.